Spec a way for a source to change the Matrix user it uses to publish information (#21)
Informo/specs: content/information-distribution/source.md

| `owner`       | `localisedString` |  x   | The company or individual maintaining this source.                                                                                                         |
| `l10n`        | `lang`            |  x   | Languages of the source's publications.                                                                                                                    |
| `sig_algo`    | `string`          |  x   | Algorithm the source will use to cryptographically sign its articles. 🔧                                                                                   |
| `sig_keys`    | `[string]`        |  x   | Public keys the source will use to cryptographically sign its articles. 🔧                                                                                 |
| `prev_id`     | `string`          |      | Matrix user ID of the Matrix user this source previously used to publish information. See [below](#change-of-matrix-user).                                 |

> | `prev_event`  | `string`          |      | ID of the latest event published by the source's previous user. Only valid if `prev_id` is set to a non-empty string. See [below](#change-of-matrix-user). |

Reviewer: ```suggestion
| `prev_event`  | `string`          |      | ID of the last legit event published by the source's previous user. Only valid if `prev_id` is set to a non-empty string. See [below](#change-of-matrix-user). |
```
There is the same change to do L161

Author: I'm not sure about the word "legit" which doesn't have an objective definition. Replacing with `the latest event published by the source with this user`.

| `owner`       | `localisedString` |  x   | The company or individual maintaining this source.                                                                                                         |
| `l10n`        | `lang`            |  x   | Languages of the source's publications.                                                                                                                    |
| `sig_algo`    | `string`          |  x   | Algorithm the source will use to cryptographically sign its articles. 🔧                                                                                   |
| `sig_keys`    | `[string]`        |  x   | Public keys the source will use to cryptographically sign its articles. 🔧                                                                                 |
| `prev_id`     | `string`          |      | Matrix user ID of the Matrix user this source previously used to publish information. See [below](#change-of-matrix-user).                                 |

> | `prev_event`  | `string`          |      | ID of the latest event published by the source's previous user. Only valid if `prev_id` is set to a non-empty string. See [below](#change-of-matrix-user). |

Reviewer: `prev_id` and `prev_event` might be better in a separated key, like `previous: {id: "@acmenews:badserver.com", event: "!xxxxx:badserver.com"}`

| `user_id`   | `string` |  x   | Matrix user ID of the Matrix user this source previously used to publish information.                                                                                                                                                                                                                                  |
| `event_id`  | `string` |  x   | ID of the latest event published by the source's original administrator using the source's previous user.                                                                                                                                                                                                              |
| `sig_algo`  | `string` |  x   | Algorithm used to generate `signature`.                                                                                                                                                                                                                                                                                |

> | `sig_key`   | `string` |  x   | Public key to use when verifying `signature`. **Must** be one of the source's public keys.                                                                                                                                                                                                                             |

Reviewer: ```suggestion
| `sig_key`   | `string` |  x   | Public key to use when verifying `signature`. **Must** be one of the source's previous user's public keys.                                                                                                                                                                                                                             |
```

| `signature` | `string` |  x   | Signature generated from a `prevUserSign` map derived from the current `prevUser` map, using the key specified in `sig_key` and the algorithm specified in `sig_algo`, and following the instructions described [here]({{<ref "/information-distribution/signature#signing-json-data">}}) (under "Signing JSON data"). |

* `prevUserSign` is a map using the following structure:

| Parameter       | Type     | Req. | Description                                                                                                                                                            |
|:----------------|:---------|:----:|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `prev_user_id`  | `string` |  x   | Matrix user ID of the Matrix user this source previously used to publish information. **Must** match the `user_id` of the current `prevUser` map.                      |

Reviewer: ```suggestion
| `prev_user_id`  | `string` |  x   | Matrix user ID this source previously used to publish information. **Must** match the `user_id` of the current `prevUser` map.                      |
```

  publication of articles in this language (and variant). This map **must**
  contain at least one element. More information on localised sub-sources and
  examples are available [below](#localisation).
* `prevUser` is a map using the following structure:

| Parameter  | Type     | Req. | Description                                                                           |
|:-----------|:---------|:----:|:--------------------------------------------------------------------------------------|
| `user_id`  | `string` |  x   | Matrix user ID of the Matrix user this source previously used to publish information. |

> | `event_id` | `string` |  x   | ID of the latest event published by the source using its previous user.               |

Reviewer: I'd say `ID of the last legit event published by the source using its previous user.`

| `event_id`  | `string` |  x   | ID of the latest event published by the source's original administrator using the source's previous user.                                                                                                                                                                                                              |
| `sig_algo`  | `string` |  x   | Algorithm used to generate `signature`.                                                                                                                                                                                                                                                                                |
| `sig_key`   | `string` |  x   | Public key to use when verifying `signature`. **Must** be one of the source's public keys.                                                                                                                                                                                                                             |

> | `signature` | `string` |  x   | Signature generated from a `prevUserSign` map derived from the current `prevUser` map, using the key specified in `sig_key` and the algorithm specified in `sig_algo`, and following the instructions described [here]({{<ref "/information-distribution/signature#signing-json-data">}}) (under "Signing JSON data"). |

Reviewer: I dont think `signature`, `sig_key`, `sig_algo` should be required, as they can be empty in case of lost sig keys

prior to the key being declared as compromised as probably not having been
tampered with, with no way to be 100% sure about it.

## Change of Matrix user

A source or subsource might have to change the Matrix user it uses to publish
information at some points in its life time. This can happen for many reason:
the homeserver it was registered on went down, or got compromised, or got
isolated from the rest of the federation...

In such an event, the source or subsource **must** publish a new registration
event from its new Matrix user with the `prev_id` key having its previous user's

> ID as its value.

Reviewer: I think you should add a matrix event ID alongside `prev_id`, in order to invalidate events from the compromised source after a given date (or event in this case), while keep using the older ones.

Reviewer: Maybe also add a reason, to let the source explain why they changed their account

Author: I added a `prev_event` key. Regarding the reason, we agreed irl it wasn't needed and would add unnecessary complexity.

This key **must** be used by client implementations to link the previous user to
the new one, considering both as the same entity. A missing `prev_id` key or an
empty string simply means that the source or subsource didn't previously use
another Matrix account to publish information.

If the entity is a source (and not a subsource), trust authorities certifying it
as trustworthy **must** update their list of trusted entities by removing the
previous user from it, and adding the new one to it. The signature associated
with the new user **must** be generated from the new user's registration event.
Client implementations **should** consider a trust authority not updating its

> list with the source's new user as if it stopped trusting it.

Reviewer: I think this is hard to enforce and potentially dangerous: If an attacker registers deceptive source with a prev_id pointing to a legit source, TAs must not update their list with the new deceptive users, and the legit source should still be trusted as if nothing happened.

The only way I see for a TA to detect if a source has changed its user, is by contacting or being contacted by the source with other communication tools.

Author: I removed lines 99 and 100.

another Matrix account to publish information. In this case, the `prev_event`
key can be omitted as well.

If the entity is a source (and not a sub-source), trust authorities certifying
it as trustworthy **must** update their list of trusted entities by removing the
previous user from it, and adding the new one to it. The signature associated
with the new user **must** be generated from the new user's registration event.

Client implementations **should** define a threshold for linking a source to its
new user. This threshold is defined by the percentage of trust authorities
certifying this source as trustworthy that updated their list of trusted sources
to reference the new user, within the trust authorities trusted by the user. If
a threshold is defined by a client implementation, it **should** consider the
link between a source and its new user as valid once that percentage becomes
higher than the threshold. Such a threshold isn't necessary for sub-sources,
because a sub-source is only verified by its parent source, which then **must**
update its list of sub-sources by removing the sub-source's previous user and
adding the new one.

> Once the link between a source (or a sub-source) and its new user is consider

Reviewer: ```suggestion
Once the link between a source (or a sub-source) and its new user is considered
```

valid, client implementations **must** consider both the source's (or
sub-source's) new user and its previous one (and older ones if the source or
sub-source changed its user more than once) as the same entity. This means that

> articles published by the source's (or the sub-source's) previous user **must**

Reviewer: `This means that
articles published by the source's (or the sub-source's) previous user before the event defined in prev_event **must**`

be treated as if it was published by the new user, and a reference to the
source's previous user in a trust authority's list of trustworthy sources
**must** be considered as a reference to the new user (with the exception of the
cryptographic check on the source's registration, which is still done using the

> registration event published by the previous user).

Reviewer: > a reference to the source's previous user in a trust authority's list of trustworthy sources **must** be considered as a reference to the new user

I don't think it's a good idea to alter the TA trusted entities list client-side:
- That would make TA appear as if they trust sources they didn't explicitly trust. That means if a client makes a wrong decision when validating the new user change, the change would then appear as 100% accepted (which is probably not the case).
- Verifying the new source's registration event against the old signature is going to fail in some cases (every time for localized sources with subsources list)

Author: ftr I let the "how the client implementation considers a link valid" (or "accepted" as you name it) unspecified on purpose to let implementations manage that the way they see fit. What is being specified here is what's happening once the client implementation decides the link is valid, either by using a threshold or having the user validate it. Either way, there's a point from which a client implementation will be able to state "now I'm pretty sure this new user is the source it pretends to be", and this paragraph specifies the UX it needs to follow from that point.

I apparently need to add some more details, as it's indeed not clear on how that impact crypto checks (which it shouldn't, the way I see it once the link is validated the "source" entity is defined by both MXIDs, and crypto checks should succeed as long as the MXID used is either the new one of an old one, and in the latter case as long as its registration has been emitted before `last_event`).

Author: Re-reading the paragraph, it already says:

> a reference to the source's previous user in a
> trust authority's list of trustworthy sources **must** be considered as a
> reference to the new user (with the exception of the cryptographic check on the
> source's registration, which is still done using the registration event
> published by the previous user).

which is pretty much what I said in my previous comment. I'll complete it to mention the cryptographic check is still done using the previous MXID, and might add a short sentence making clear that the source, as an entity, is defined by all of the MXIDs it's used, but otherwise there's not much to add.

Client implementations **can** warn users reading articles published by a
source's or a sub-source's previous user that this user isn't active anymore,
and might get compromised at some point. This is the only exception to the equal
treatment principle mentioned in the previous paragraph.

{{% notice tip %}}

> If a trust authority certifying the source or subsource that's chaning its

Reviewer: ```suggestion
If a trust authority certifying the source or subsource that's changing its
```

information at some points in its life time. This can happen for many reason:
the homeserver it was registered on went down, or got compromised, or got
isolated from the rest of the federation...

In such an event, the source or sub-source **must** publish a new registration
event from its new Matrix user with the `prev_id` key having its previous user's
ID as its value, and the `prev_event` key having the ID of the latest event
emitted prior to the change as its value.

This key **must** be used by client implementations to link the previous user to
the new one, considering both as the same entity. A missing `prev_id` key or an
empty string simply means that the source or sub-source didn't previously use
another Matrix account to publish information. In this case, the `prev_event`
key can be omitted as well.

If the entity is a source (and not a sub-source), trust authorities certifying
it as trustworthy **must** update their list of trusted entities by removing the
previous user from it, and adding the new one to it. The signature associated
with the new user **must** be generated from the new user's registration event.

> Client implementations **should** define a threshold for linking a source to its

Reviewer: I'd say "can", since very suspicious/protective clients should not implement any threshold. It's just a comfort feature, in order to require less manual validations.

I can also think of another client implementation that would have two "levels of trusting" the user can chose when trusting a TA: "ultimate" would synchronize TA properties/user changes with the client storage silently, while "normal" would require either a custom threshold (trusted by one ultimate TA, or "x" normal TAs, or "y" percent of normal TAs) or manual validation in order to synchronize the changes.

Reviewer: IMHO it's the client developers' role to decide how to balance usability versus security, not the spec's role. The threshold thing is more a client UX design implementation issue.
We'll probably need to discuss about this on voice chat :)

Author: Makes sense.

prior to the key being declared as compromised as probably not having been
tampered with, with no way to be 100% sure about it.

## Change of Matrix user

A source or sub-source might have to change the Matrix user it uses to publish
information at some points in its life time. This can happen for many reason:
the homeserver it was registered on went down, or got compromised, or got
isolated from the rest of the federation...

In such an event, the source or sub-source **must** publish a new registration
event from its new Matrix user with the `prev_user` key set. The value of this
key **must** be a map in which the `user_id` key's value is the source's or
sub-source's previous user's, and the `event_id` key's value is the ID of the
latest event emitted prior to the change.

This key **must** be used by client implementations to link the previous user to
the new one, considering both as the same entity. A missing `prev_user` key
simply means that the source or sub-source didn't previously use another Matrix
account to publish information.

If the entity is a source (and not a sub-source), trust authorities certifying

> it as trustworthy **must** update their list of trusted entities by removing the

Reviewer: ```suggestion
it as trustworthy **must** verify the new user authenticity and update their list of trusted entities by removing the
```
maybe add a note saying they should contact the trusted user through other communication mediums? (just for the sake of being very explicit)

previous user from it, and adding the new one to it. The signature associated
with the new user **must** be generated from the new user's registration event.

### Linking between a source and its new user

> Client implementations are in charge of asserting the vailidy of a link between

Reviewer: ```suggestion
Client implementations are in charge of asserting the validity of a link between
```

prior to the key being declared as compromised as probably not having been
tampered with, with no way to be 100% sure about it.

## Change of Matrix user

A source or sub-source might have to change the Matrix user it uses to publish
information at some points in its life time. This can happen for many reason:
the homeserver it was registered on went down, or got compromised, or got
isolated from the rest of the federation...

In such an event, the source or sub-source **must** publish a new registration
event from its new Matrix user with the `prev_user` key set. The value of this
key **must** be a map in which the `user_id` key's value is the source's or
sub-source's previous user's, and the `event_id` key's value is the ID of the
latest event emitted prior to the change.

> This key **must** be used by client implementations to binding the previous user

Reviewer: ```suggestion
This key **must** be used by client implementations to bind the previous user
```

to the new one, considering both as the same entity. A missing `prev_user` key
simply means that the source or sub-source didn't previously use another Matrix
account to publish information.

If the entity is a source (and not a sub-source), trust authorities certifying
it as trustworthy **must** verify that the change comes from the source's
original administrator (preferably using a different medium than Informo) and
update their list of trusted entities by removing the previous user from it, and
adding the new one to it. The signature associated with the new user **must** be
generated from the new user's registration event.

### Binding between a source and its new user

If the `signature` key from the `prev_user` key in the new user's registration
event can be verified using a public cryptographic key that was already in the
previous user's registration event's `sig_keys` key, then client implementations
**should** consider the binding between a source (or a sub-source) and its new

> user valid.

Reviewer: ```suggestion
user to be valid.
```
I find it easier to read like this

the homeserver it was registered on went down, or got compromised, or got
isolated from the rest of the federation...

In such an event, the source or sub-source **must** publish a new registration
event from its new Matrix user with the `prev_user` key set. The value of this
key **must** be a map in which the `user_id` key's value is the source's or
sub-source's previous user's, and the `event_id` key's value is the ID of the
latest event emitted prior to the change.

This key **must** be used by client implementations to link the previous user to
the new one, considering both as the same entity. A missing `prev_user` key
simply means that the source or sub-source didn't previously use another Matrix
account to publish information.

If the entity is a source (and not a sub-source), trust authorities certifying
it as trustworthy **must** verify that the change comes from the source's
original administrator (preferably using a different medium than Informo) and
update their list of trusted entities by removing the previous user from it, and
adding the new one to it. The signature associated with the new user **must** be
generated from the new user's registration event.

> ### Linking between a source and its new user

Reviewer: I just realized we use the word "link" for two different associations:
- links between a TA and other entities, representing a trust relationship
- links from a previous mxid to a new mxid, representing a "technical user migration"

Maybe we should consider changing this kind of links to something like "migration", "migration link", "user transfer", "user bond", ...

Please tell me your thoughts about this

Author: Using "binding" for that.

> If the new user's registration event is signed by a cryptographic key it was

Reviewer: I think you forgot to remove the first sentense ;)

using before the If the `signature` key from the `prev_user` key in the new
user's registration event can be verified using a public cryptographic key that
was already in the previous user's registration event's `sig_keys` key, then

> client implementations **must** consider the link between a source (or a

Reviewer: ```suggestion
client implementations **should** consider the link between a source (or a
```
Security-focused clients should be allowed to ask users to confirm mxid changes

sub-source) and its new user valid.

If such a condition couldn't be met, client implementations are in charge of
asserting the validity of a link between a source, or sub-source, and its new
user, using whatever mechanism they see fit.

{{% notice tip %}}
A suggested way to assert the validity of this link would be for a client
implementation to define a threshold for linking a source to its new user. This
threshold is defined by the percentage of trust authorities certifying this
source as trustworthy that updated their list of trusted sources to reference
the new user, within the trust authorities trusted by the user. If a threshold
is defined by a client implementation, it should consider the link between a
source and its new user as valid once that percentage becomes higher than the
threshold. Such a threshold wouldn't be necessary for sub-sources, because a
sub-source is only verified by its parent source, which then needs to update its
list of sub-sources by removing the sub-source's previous user and adding the

> new one.

Reviewer: You know I don't like the threshold idea ^^

IMHO automating (or giving hints for) user migration validation based on any kind of heuristics/approximation can cause a lot of security issues.

If you think this example is important, I'd like you to add a warning like:
`However, client implementations must be aware that such approximations may be exploited and cause a deceptive source becomes trusted.`

{{% /notice %}}

Once the link between a source (or a sub-source) and its new user is considered
valid, client implementations **must** consider both the source's (or
sub-source's) new user and its previous one (and older ones if the source or
sub-source changed its user more than once) as the same entity (i.e. as an
entity defined by both its previous users' IDs and its new one). This means that
articles published by the source's (or the sub-source's) previous user before
the event defined in the `event_id` key **must** be treated as if it was
published by the new user, and a reference to the source's previous user in a

> trust authority's list of trustworthy sources **must** be considered as a

Reviewer: I don't think the new user should replace the older one in a TA's list

| Parameter     | Type       | Req. | Description                                                                                                                                                    |
|:--------------|:-----------|:----:|:---------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `parent`      | `string`   |  x   | Matrix user ID of the sub-source's parent.                                                                                                                     |
| `sig_algo`    | `string`   |  x   | Algorithm the sub-source will use to cryptographically sign its articles. 🔧                                                                                   |
| `sig_keys`    | `[string]` |  x   | Public keys the sub-source will use to cryptographically sign its articles. 🔧                                                                                 |
| `prev_id`     | `string`   |      | Matrix user ID of the Matrix user this sub-source previously used to publish information. See [above](#change-of-matrix-user).                                 |

> | `prev_id`     | `string`   |      | ID of the latest event published by the sub-source's previous user. Only valid if `prev_id` is set to a non-empty string. See [below](#change-of-matrix-user). |

Reviewer: ```suggestion
| `prev_event`     | `string`   |      | ID of the latest event published by the sub-source's previous user. Only valid if `prev_id` is set to a non-empty string. See [below](#change-of-matrix-user). |
```
copy pasta spotted :)

    "owner": {
        "en": "ACME News Group"
    },

> 	"prev_id": "@acmenews:badserver.com",

Reviewer: There's a tab character here instead of spaces, and at other places too

Author: Fixed (everywhere, hopefully)
